tools/expotools/src/commands/ClientBuild.ts

```diff
@@ -1,144 +1,177 @@
+  }
+
+  const builder = getBuilderForPlatform(platform);
+  const sdkVersion = sdkBranchVersion || (await askForSdkVersionAsync(platform));
+  const appVersion = await builder.getAppVersionAsync();
+
+  await buildOrUseCacheAsync(builder);
+
+  if (sdkVersion && !options.skipUpload) {
+    await uploadAsync(builder, sdkVersion, appVersion);
+  }
+  if (sdkVersion && options.release) {
+    await releaseAsync(builder, sdkVersion, appVersion);
+  }
+}
+
+function getBuilderForPlatform(platform: Platform): ClientBuilder {
+  switch (platform) {
+    case 'ios':
+      return new IosClientBuilder();
+    case 'android':
+      return new AndroidClientBuilder();
+    default: {
+      throw new Error(`Platform "${platform}" is not supported yet!`);
+    }
+  }
+}
+
+async function askToRecreateSimulatorBuildAsync(): Promise<boolean> {
   if (process.env.CI) {
     return false;
   }
   const { createNew } = await inquirer.prompt<{ createNew: boolean }>([
     {
       type: 'confirm',
       name: 'createNew',
-      message: `Simulator archive already exists at ${chalk.magenta(
-        path.relative(EXPO_DIR, archivePath)
-      )}. Do you want to create a fresh one?`,
+      message: 'Do you want to create a fresh one?',
       default: true,
     },
   ]);
   return createNew;
 }
 
-function getApplicationPathForPlatform(platform: Platform) {
-  switch (platform) {
-    case 'ios': {
-      return path.join(
-        IOS_DIR,
-        'simulator-build',
-        'Build',
-        'Products',
-        'Release-iphonesimulator',
-        'Exponent.app'
-      );
-    }
-    case 'android': {
-      return path.join(ANDROID_DIR, 'app', 'build', 'outputs', 'apk', 'release', 'app-release.apk');
-    }
-    default: {
-      throw new Error(`Platform "${platform}" is not supported yet!`);
-    }
+async function askToOverrideBuildAsync(): Promise<boolean> {
+  if (process.env.CI) {
+    return true;
   }
+  const { override } = await inquirer.prompt<{ override: boolean }>([
+    {
+      type: 'confirm',
+      name: 'override',
+      message: 'Do you want to override it?',
+      default: true,
+    },
+  ]);
+  return override;
 }
 
-async function buildAndReleaseClientAsync(
-  platform: Platform,
-  sdkVersion: string | undefined,
-  release: boolean
-) {
-  const appPath = getApplicationPathForPlatform(platform);
-
-  if (!(await fs.pathExists(appPath)) || (await askToRecreateSimulatorBuildAsync(appPath))) {
-    const args =
-      platform === 'ios'
-        ? ['ios', 'create_simulator_build']
-        : ['android', 'build', 'build_type:Release'];
-
-    await spawnAsync('fastlane', args, {
-      cwd: EXPO_DIR,
-      stdio: 'inherit',
-      env: { ...process.env },
-    });
-  } else {
-    console.log(`Using cached build at ${chalk.magenta(path.relative(EXPO_DIR, appPath))}...`);
-  }
+async function buildOrUseCacheAsync(builder: ClientBuilder): Promise<void> {
```

Comment:
I know that these scripts are rather expected to simple, but I wonder if `caching` concept can be somehow extracted and shared across whole `expotools` command?
I wonder if `Builder`s classes should know about `caching` as well (and thus expose method `buildOrUseCacheAsync` by themselves? 🤔) or rather it actually should be written as you;re proposing right here 🤔 
From my point of view `caching` concept might come in handy in many places and maybe it's worth to prepare some general technique for adding it by simply `wrapping` or `decorating` original methods/classes? What would you say?

Reply:
Generally I don't think we really need caching here. We needed it in CircleCI because one job was building and the other one was releasing (after approval), but afaik on GitHub Actions it is different now. Maybe you're right with the idea, but it's also not in scope of this PR 😛